reference-implementation/lib/readable-stream.js

```diff
@@ -86,273 +89,126 @@ class ReadableStream {
+          performShutdown(true, storedError);
+        }
+      });
+
+      // Errors must be propagated backward
+      writer._closedPromise.catch(storedError => {
+        if (preventCancel === false) {
+          performAsyncShutdown(() => ReadableStreamCancel(this, storedError), true, storedError);
+        } else {
+          performShutdown(true, storedError);
+        }
+      });
+
+      // Closing must be propagated forward
+      reader._closedPromise.then(() => {
+        if (preventClose === false) {
+          performAsyncShutdown(() => WritableStreamDefaultWriterCloseWithErrorPropagation(writer));
+        } else {
+          performShutdown();
+        }
+      });
 
-        // We don't use writer.closed. We can ensure that the microtask for writer.closed is run before any
-        // writer.close() call so that we can determine whether the closure was caused by the close() or ws was already
-        // closed before pipeTo(). It's possible but fragile.
-        _writer.close().then(
-          () => {
-            return _allWrites;
-          },
-          reason => {
-            releaseWriter();
-            finishWithRejection(reason);
-          }
-        ).then(
-          () => {
-            releaseWriter();
-            finishWithFulfillment();
-          }
-        );
-        _state = 'closingDest';
+      // Closing must be propagated backward
+      writer._closedPromise.then(() => {
+        const destClosed = new TypeError('the destination writable stream closed before all data could be piped to it');
 
-        return;
-      }
+        if (preventCancel === false) {
+          performAsyncShutdown(() => ReadableStreamCancel(this, destClosed), true, destClosed);
+        } else {
+          performShutdown(true, destClosed);
+        }
+      });
 
-      if (_lastWrite === undefined) {
-        releaseWriter();
-        finishWithFulfillment();
-        return;
+      function waitForCurrentWrite() {
+        return currentWrite.catch(() => {});
       }
 
-      // We don't use writer.closed. pipeTo() is responsible only for what it has written.
-      _lastWrite.then(
-        () => {
-          releaseWriter();
-          finishWithFulfillment();
-        },
-        reason => {
-          releaseWriter();
-          finishWithRejection(reason);
+      function performAsyncShutdown(action, originalIsError, originalError) {
+        if (shuttingDown === true) {
+          return;
         }
-      );
-      _state = 'waitingLastWriteOnReadableClosed';
-    }
-
-    function doPipe() {
-      // console.log('pipeTo(): doPipe()');
+        shuttingDown = true;
+
+        waitForCurrentWrite().then(() => {
+          action().then(
+            () => performShutdown(originalIsError, originalError),
+            newError => performShutdown(true, newError)
+          );
+        });
+      }
 
-      _lastRead = _reader.read();
+      function performShutdown(isError, error) {
+        shuttingDown = true;
```

Comment:
_shuttingDown_ might be true here, if we are coming from async shutdown. So that would be expected, not surprising.

_shuttingDown_ is basically meant to prevent a second async shutdown from happening if one is already in progress.